### DISCUSSION TITLE
Sending event.programChange adds +1 to "program" value

### DISCUSSION THREAD

**Discussion by NikolaDeveloper** [orchetect/swift-midi#181] [Q&A] [answered]
I managed to make simple demo for changing programs on my synth but I'm having trouble sending program with value of 0 (it always adds +1 to whatever I put in "program" argument). Here's example:

`try? outputConnection?.send(event: .programChange(
    program: 0,
    bank: .bankSelect(msb: 64, lsb: 32),
    channel: 0
))`

This sends the following (captured with MIDI Monitor):

```
16:23:37.845	From iPad	Control	1	Bank Select	64
16:23:37.845	From iPad	Control	1	Bank Select (fine)	32
16:23:37.845	From iPad	Program	1	1
```

Some MIDI systems have value ranges from 0-127 while others have 1-128, and I assume MIDIKit is written for 1-128 systems so I was wondering if there's an option I could toggle so it doesn't add +1 to program value?

**Comment by orchetect** [comment, marked as answer]
Program numbers are always encoded as 7-bit values which are integers 0-127. There is no addition happening. There are not different "systems". What you're describing is encoding vs. presentation. MIDI Spec designates program changes to be encoded 0-127 (zero-based) but when presented to the user they are 1-128 (1-based). Your logging software is displaying its presentation number, not its encoding value.

<img width="679" alt="midi-spec" src="https://user-images.githubusercontent.com/11605557/227008362-0fdb45d1-a90d-4a63-bf92-6c357b0a2bb0.png">

Similar to how channels are encoded 0-15 but MIDI software displays them as 1-16.

Not all values in the MIDI Spec work this way (for example, CC controller numbers and note velocity are both encoded 0-127 but are also presented to the user as 0-127).

**Reply by NikolaDeveloper** [reply]
That does actually make perfect sense, my bad for thinking MIDI Monitor would show encoding as it is. Thank you for this amazing library, made me actually start learning ios development!
